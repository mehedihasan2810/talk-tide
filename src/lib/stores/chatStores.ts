import { create } from "zustand";

interface State {
  isGroupChat: boolean;
  groupName: string;
  selectedUser: string | null;
  groupParticipants: string[];
}

interface Action {
  updateIsGroupChat(_a: boolean): void;
  updateGroupName(_a: string): void;
  updateSelectedUser(_a: string): void;
  updateGroupParticipants(_a: string): void;
}

export const useChatStore = create<State & Action>((set) => ({
  // STATE
  isGroupChat: false,
  groupName: "",
  selectedUser: null,
  groupParticipants: [],

  // ACTION
  updateIsGroupChat: (isGroupChat) => set(() => ({ isGroupChat })),
  updateGroupName: (groupName) => set(() => ({ groupName })),
  updateSelectedUser: (selectedUser) => set(() => ({ selectedUser })),
  updateGroupParticipants: (groupParticipant) =>
    set((state: State) => {
      const participants = state.groupParticipants.includes(groupParticipant)
        ? state.groupParticipants
        : [...state.groupParticipants, groupParticipant];

      return { groupParticipants: participants };
    }),
}));
