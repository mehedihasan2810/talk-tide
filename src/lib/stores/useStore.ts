import { create } from "zustand";

interface ChatStore {
  isMobileSidebarOpen: boolean;
  isGroupDetailsModalOpen: boolean;

  toggleIsMobileSidebarOpen(): void;
  toggleGroupDetailsModal(_isGroupDetailsModalOpen: boolean): void;
}

export const useStore = create<ChatStore>((set) => ({
  // STATE
  isMobileSidebarOpen: false,
  isGroupDetailsModalOpen: false,
  // --------------------------

  // ACTION
  toggleIsMobileSidebarOpen: () =>
    set(({ isMobileSidebarOpen }) => ({
      isMobileSidebarOpen: !isMobileSidebarOpen,
    })),

  toggleGroupDetailsModal: (isGroupDetailsModalOpen) =>
    set({ isGroupDetailsModalOpen }),
}));
