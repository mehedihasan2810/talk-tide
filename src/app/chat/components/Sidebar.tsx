import { useChatStore } from "@/lib/stores/chatStores";
import LargeSidebar from "./LargeSidebar";
import MobileSidebar from "./MobileSidebar";

const Sidebar = () => {
  const toggleIsMobileSidebarOpen = useChatStore(
    (state) => state.toggleIsMobileSidebarOpen
  );

  return (
    <>
      {/* ------------ */}
      <button
        onClick={toggleIsMobileSidebarOpen}
        className="hidden"
        data-testid="close-mobile-sidebar-btn"
      ></button>

      {/* ---------------- */}
      <div className="md:hidden">
        <MobileSidebar />
      </div>
      <div className="hidden md:block">
        <LargeSidebar />
      </div>
    </>
  );
};

export default Sidebar;
