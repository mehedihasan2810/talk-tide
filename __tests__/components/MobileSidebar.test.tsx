import Sidebar from "@/app/chat/components/Sidebar";
import { useStore } from "@/lib/stores/useStore";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("<MobileSidebar/>", () => {
  it("should render successfully and match snapshot", async () => {
    const { asFragment } = render(<Sidebar />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("sidebar should toggle perfectly", async () => {
    // UTILS
    const user = userEvent.setup();

    const toggleSidebar = jest.spyOn(
      useStore.getState(),
      "toggleIsMobileSidebarOpen",
    );
    // ---------------------------

    render(<Sidebar />); //ARRANGE

    // ASSERT
    expect(screen.queryByTestId("sidebar-dialog")).not.toBeInTheDocument();
    expect(toggleSidebar).not.toHaveBeenCalled();

    // ACT
    const closeBtn = screen.getByTestId("close-mobile-sidebar-btn");
    await user.click(closeBtn);

    // ASSERT
    expect(toggleSidebar).toHaveBeenCalled();
    expect(screen.getByTestId("sidebar-dialog")).toHaveAttribute(
      "data-state",
      "open",
    );
    expect(closeBtn).toBeInTheDocument();
  });
});
