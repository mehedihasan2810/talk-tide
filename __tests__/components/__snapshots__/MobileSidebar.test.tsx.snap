// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MobileSidebar/> should render successfully and match snapshot 1`] = `
<DocumentFragment>
  <button
    class="hidden"
    data-testid="close-mobile-sidebar-btn"
  />
  <div
    class="md:hidden"
  />
  <div
    class="hidden md:block"
  >
    <div>
      <div
        class="border-b mb-4 h-14 px-4 flex justify-between items-center"
      >
        <p>
          hello my name is
        </p>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground w-8 h-8 rounded-full p-1"
        >
          <svg
            aria-hidden="true"
            class="w-full h-full text-gray-700"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M7.84 1.804A1 1 0 018.82 1h2.36a1 1 0 01.98.804l.331 1.652a6.993 6.993 0 011.929 1.115l1.598-.54a1 1 0 011.186.447l1.18 2.044a1 1 0 01-.205 1.251l-1.267 1.113a7.047 7.047 0 010 2.228l1.267 1.113a1 1 0 01.206 1.25l-1.18 2.045a1 1 0 01-1.187.447l-1.598-.54a6.993 6.993 0 01-1.929 1.115l-.33 1.652a1 1 0 01-.98.804H8.82a1 1 0 01-.98-.804l-.331-1.652a6.993 6.993 0 01-1.929-1.115l-1.598.54a1 1 0 01-1.186-.447l-1.18-2.044a1 1 0 01.205-1.251l1.267-1.114a7.05 7.05 0 010-2.227L1.821 7.773a1 1 0 01-.206-1.25l1.18-2.045a1 1 0 011.187-.447l1.598.54A6.993 6.993 0 017.51 3.456l.33-1.652zM10 13a3 3 0 100-6 3 3 0 000 6z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
      <div
        class="px-2"
      >
        <div
          class="flex items-center space-x-2 mb-2"
        >
          <button
            aria-checked="false"
            class="peer inline-flex h-[24px] w-[44px] shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=unchecked]:bg-input"
            data-state="unchecked"
            id="group-chat"
            role="switch"
            type="button"
            value="on"
          >
            <span
              class="pointer-events-none block h-5 w-5 rounded-full bg-background shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-5 data-[state=unchecked]:translate-x-0"
              data-state="unchecked"
            />
          </button>
          <label
            class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
            for="group-chat"
          >
            Is it a group chat?
          </label>
        </div>
        <div
          class="flex flex-wrap gap-1  mb-2"
        />
        <div
          class="flex gap-2"
        >
          <button
            aria-controls="radix-:r3:"
            aria-expanded="false"
            aria-haspopup="dialog"
            class="inline-flex items-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 w-full justify-between border-gray-300"
            data-state="closed"
            role="combobox"
            type="button"
          >
            Select User...
            <svg
              aria-hidden="true"
              class="ml-2 h-4 w-4 shrink-0 opacity-50"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M10 3a.75.75 0 01.55.24l3.25 3.5a.75.75 0 11-1.1 1.02L10 4.852 7.3 7.76a.75.75 0 01-1.1-1.02l3.25-3.5A.75.75 0 0110 3zm-3.76 9.2a.75.75 0 011.06.04l2.7 2.908 2.7-2.908a.75.75 0 111.1 1.02l-3.25 3.5a.75.75 0 01-1.1 0l-3.25-3.5a.75.75 0 01.04-1.06z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 py-2 px-3 border-gray-300"
          >
            Start
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
